src/core/border_router/routing_manager.hpp

```diff
@@ -953,18 +953,60 @@ class RoutingManager : public InstanceLocator
         ExpireTimer                       mTimer;
     };
 
-    typedef Ip6::Prefix OnMeshPrefix;
+    void HandleRioAdvertiserimer(void) { mRioAdvertiser.HandleTimer(); }
 
-    class OnMeshPrefixArray :
+    class RioAdvertiser : public InstanceLocator
+    {
+        // Manages the list of prefixes advertised as RIO in emitted
+        // RA. The RIO prefixes are discovered from on-mesh prefixes in
+        // network data including OMR prefix from `OmrPrefixManager`.
+        // It also handles deprecating removed prefixes.
+
+    public:
+        explicit RioAdvertiser(Instance &aInstance);
+
+        RoutePreference GetPreference(void) const { return mPreference; }
+        void            SetPreference(RoutePreference aPreference);
+        void            ClearPreference(void);
+        void            HandleRoleChanged(void);
+        void            AppendRios(Ip6::Nd::RouterAdvertMessage &aRaMessage);
+        void            InvalidatPrevRios(Ip6::Nd::RouterAdvertMessage &aRaMessage);
+        bool            HasAdvertised(const Ip6::Prefix &aPrefix) const { return mPrefixes.ContainsMatching(aPrefix); }
+        uint16_t        GetAdvertisedRioCount(void) const { return mPrefixes.GetLength(); }
+        void            HandleTimer(void);
+
+    private:
+        static constexpr uint32_t kDeprecationTime = TimeMilli::SecToMsec(300);
+
+        struct RioPrefix : public Clearable<RioPrefix>
+        {
+            bool Matches(const Ip6::Prefix &aPrefix) const { return (mPrefix == aPrefix); }
+
+            Ip6::Prefix mPrefix;
+            bool        mIsDeprecating;
+            TimeMilli   mExpirationTime;
+        };
+
+        struct RioPrefixArray :
 #if OPENTHREAD_CONFIG_BORDER_ROUTING_USE_HEAP_ENABLE
-        public Heap::Array<OnMeshPrefix>
+            public Heap::Array<RioPrefix>
 #else
-        public Array<OnMeshPrefix, kMaxOnMeshPrefixes>
+            public Array<RioPrefix, 2 * kMaxOnMeshPrefixes>
```

Comment:
Earlier there was a factor 3 used (in `kMaxRaLength` definition). Is 2 ok here?

Reply:
Yes. The 3 above (for RA message buffer) is to account for prefixes that are fully removed. We include them as RIO with zero lifetime in the RA message. We don't need to track the removed prefixes anymore. Here in this array we are tracking all active and deprecating RIO prefixes that were advertised in the last emitted RA, so factor 2 is good.

I recently added option to use heap instead which is more flexible and is now the default config (still kept the original model of fixed array in case it is needed/used by some platforms).